### DISCUSSION TITLE
Can maplibre handle sparse pmtiles?

### DISCUSSION THREAD

**Discussion by etienneJr** [maplibre/maplibre-gl-js#5126] [Q&A]
Hi,
I don't have the disk space on my server to host a pmtiles file with all levels for the whole planet. I was thinking of preparing a sparse pmtiles with :
- the whole planet up to zoom (let's say) 10
- my country up to zoom 14.

I'am expecting the following behaviour for levels 11 to 14:
- in my country it displays the available tiles
- elsewhere in the world, it overzooms in the level 10 tiles (the last level available)

I've done some preliminary tests on small regions and it doesn't seem to work:
- I have to indicate in the pmtiles metadata that the levels go from 0 to 14 otherwise it won't look for tiles at levels 11-14 when zooming in my country 
- and so, elsewhere in the world, it looks for tiles from levels 11 to 14, which don't exist, so it doesn't display anything anymore, instead of overzooming in the last available level (level 10).

Is it possible to handle sparse pmtiles with overzooming in the last available level ? Is it a question for maplibre developers or pmtiles developers ?
Thanks !
Etienne

**Comment by HarelM** [comment]
I believe the solution would involve 2 pmtiles behind a tile server and routing though nginx rules. I haven't tried that, but I think you should be able to make it work...

**Reply by wipfli** [reply]
Or you could maybe also do it with addProtocol where you route the request to the correct pmtiles file. I think I did this once with raster tiles on a z/x/y endpoint but I have not done it with pmtiles. Is it possible to chain multiple addProtocol calls @HarelM ?

**Reply by HarelM** [reply]
No, unfortunately you can't chain multiple addProtocol...

**Reply by etienneJr** [reply]
yes I had thought of this solution with 2 files, and I'll do that for my current case if their is no other solution.
But a solution with only 1 file supported by maplibre would be easier to deploy by other people who don't have the skills or control over their server.
Is this a possible feature request or is it technically impossible to ask for overzooming in previous level tile when actual level tile is not available ?

**Reply by HarelM** [reply]
I think this is the default behavior as far as I remember for 404 response, but I'm not 100% sure.

**Reply by etienneJr** [reply]
I checked in the browser's dev console, and (using the pmtiles protocol) I do not get a 404 response or any other error if the tile is not found.
So I checked the protocol in the [pmtiles.js file](https://unpkg.com/browse/pmtiles@3.0.6/dist/pmtiles.js) I'm using (it's v3.0.6 but I checked with v4.0.1 and got the same behaviour), and it seems to me that the answer is an empty array when the tile is not found:

```
var Protocol = class {
  constructor() {
    this.tilev4 = (params, abortController) => __async(this, null, function* () {
      (...)
      const z = result[2];
      const x = result[3];
      const y = result[4];
      const header = yield instance.getHeader();
      const resp = yield instance == null ? void 0 : instance.getZxy(+z, +x, +y, abortController.signal);
      // .getZxy() RETURNS UNDEFINED IF TILE NOT FOUND
      if (resp) {
        //WHEN TILE IS FOUND
        return {
          data: new Uint8Array(resp.data),
          cacheControl: resp.cacheControl,
          expires: resp.expires
        };
      }
      if (header.tileType === 1 /* Mvt */) {
        //WHEN TILE IS NOT FOUND
        return { data: new Uint8Array() };
      }
      return { data: null };
    });
 (...)
```
So you think that if the pmtiles protocol is changed to send a 404 error, maplibre will overzoom in the previous level?

**Reply by HarelM** [reply]
It's worth trying to see what happens.

**Reply by etienneJr** [reply]
OK, I patched the pmtiles protocol to throw a 404 error instead of an empty array, and it works just as expected : z12 tiles are overzoomed by maplibre when z13 or z14 tiles return errors. You can see more details [in this discussion](https://github.com/protomaps/PMTiles/discussions/321#discussioncomment-11501808).
Now, how can we make this behaviour accessible to other people who would like to use sparse pmtiles?
Can I ask maplibre developers to include an additional parameter for overzooming when the tile is empty, not only when there is a 404 response? or would it not be compatible with other maplibre functions?

**Reply by HarelM** [reply]
Empty content in http is marked with 204 status code.
I'm not sure what was the last decision around this, see here:
https://github.com/maplibre/maplibre-gl-js/pull/3428

It would be interesting to see if this is something you can configure for pmtiles library, as the return of empty tile is "hard coded" choice and it might make sense to allow defining a different behavior in certain cases.

Generally speaking, an empty response is a valid one, and I'm not sure we want to start hacking the code in order to change the behavior of empty tiles...

**Reply by etienneJr** [reply]
I agree that an empty response is a valid one, and I'd rather keep it that way than throw in a lot of errors. However, from a graphic designer's perspective, a transparent layer means "show me what's below". When the user has several sources, it is logical that _below_ refers to another source. But when the user has only 1 source, it is logical that _below_ refers to the previous level in that source. So to my point of view, when the user defines a source, he should be able to choose maplibre behaviour when handling empty tiles : either display nothing, or overzoom the previous level in that source. We could imagine for example an _empty-tile-handling_ parameter in the source definition.
Obviously, I'm saying this without having any idea of the complexity of all the cases that maplibre must be able to manage.

**Reply by HarelM** [reply]
You are welcome to open a design proposal in the maplibre-style-spec repo and proceed with the relevant implementation if this is something that you would like to push forward.

**Reply by etienneJr** [reply]
OK I 'll do that !

**Comment by acalcutt** [comment]
I have not tested this with vector tiles, but I have been looking at something similar with terrain raster-dem files. I talked with wipfli initially and he had given me some ideas for layering raster tiles with addprotocol that replaced specific tiles.  That worked, but it wasn't very flexible since you had to specify every tile you wanted to replace at every zoom level.

I took that example and asked google AI to make a function that would check if a higher zoom tile was withing a lower zoon tile, and it gave me the function I used in this example. I used that function to show a different source when the tile is within the tiles (5/16/11, 5/17/11, 5/16/12, 5/17/12)
https://stackblitz.com/edit/web-platform-j8cf5x?file=index.html

In my example the main source goes to zoom 12, but the source in the  (5/16/11, 5/17/11, 5/16/12, 5/17/12) areas go up to zoom 14. For the most part it seems with terrain that it does try to load lower zoom tiles if the higher zoom tiles don't exist. However even though the terrain works, I think I have also run into some issues caused tile zoom level being different from the max zoom on the source (specifically 'merged_terrain_source' in my example). For example, When I use it with a style that includes buildings and am looking in an area that only goes to zoom 12 instead of the source max zoom of 14, it seems to cause the building to float (which does not happen in the areas that actually go to zoom 14)
https://stackblitz.com/edit/web-platform-mkbj4hau?file=index.html
![image](https://github.com/user-attachments/assets/becd5489-fb93-4050-9913-533fa3239671)
this does not seem to happen in areas that actually go to zoom 14 ([for example](https://wifidb.net/demo/terrain_layer/#map=16.94/47.285288/7.716903/-8/67))

It also goes away if i set the 'merged_terrain_source'  to max zoom of 12, but then it wouldn't load the higher zoom tiles where they are available.

**Reply by acalcutt** [reply]
in testing I looked at the developer tools network tab and can see it loads lower zoom tiles. like this for example, you can see it failed to load higher zoom images (which were at zoom 14 and 13) and then successfully loads a zoom 12 tile
![image](https://github.com/user-attachments/assets/a8063de8-c2f4-4954-acbf-9a750949156b)

I would imagine you would see something similar with pbf files if it was working.

**Comment by spatialillusions** [comment]
For PMTiles I think the solution is to initiate with the following code like I do in #5618 

```js
new pmtiles.Protocoll({
   errorOnMissingTile: true,
});
```

It is not an ideal solution, but it sort of works.

**Comment by zstadler** [comment]
I'm unable to reproduce the over-zooming behavior described above when using a remote vector tile server. The map turns blank when a 404 response arrives.
On the right side of the screenshot is a tile that does not exist at zoom=14
<img width="487" height="400" alt="image" src="https://github.com/user-attachments/assets/d6e9db30-6ab9-49eb-9582-517a2571a0dd" />

**Comment by HarelM** [comment]
The following PR might be interesting for this discussion:
- #8335
